### ISSUE TITLE
socket.connect fatally errors on unset options

### ISSUE THREAD

**Issue by AlissaSquared** [wahern/cqueues#77]
My kernel does weird things with libc in that if I try to set off `SO_REUSEPORT`, it'll still error. cqueues fully errors on this. Is there a way we can get it so cqueues doesn't fatally error?

Significant with [this paste](http://phosphor.i0i0.me/Ote0z0td) and `SO_REUSEPORT`.

tl;dr: only error with .connect() if trying to turn on a value if option isn't supported, not including setting a value off.

**Comment by daurnimator** [comment]
> My kernel does weird things with libc in that if I try to set off SO_REUSEPORT, it'll still error. cqueues fully errors on this. Is there a way we can get it so cqueues doesn't fatally error?

This patch should fix it:

```diff
diff --git a/src/lib/socket.c b/src/lib/socket.c
index 984b7ef..3206dae 100644
--- a/src/lib/socket.c
+++ b/src/lib/socket.c
@@ -807,12 +807,14 @@ int so_reuseaddr(int fd, _Bool reuseaddr) {
 
 
 int so_reuseport(int fd, _Bool reuseport) {
+	int err;
 #if defined SO_REUSEPORT
-	return so_setboolopt(fd, SOL_SOCKET, SO_REUSEPORT, reuseport);
+	err = so_setboolopt(fd, SOL_SOCKET, SO_REUSEPORT, reuseport);
 #else
 	(void)fd;
-	return (reuseport)? EOPNOTSUPP : 0 /* already disabled */;
+	err = EOPNOTSUPP;
 #endif
+	return (err != EOPNOTSUPP || reuseport)? err : 0 /* already disabled */;
 } /* so_reuseport() */

```

**Comment by daurnimator** [comment]
Closed with 852dd24648def2fb0cad0a1891951f30c9247b9f
